libgalois/src/analytics/cdlp/cdlp.cpp

```diff
@@ -0,0 +1,302 @@
+      // No change! break!
+      if (applyBag.empty())
+        break;
+
+      // Apply Phase
+      katana::do_all(
+          katana::iterate(applyBag),
+          [&](const NodeDataPair nodeData) {
+            GNode node = nodeData.node;
+            auto& ndata = graph->GetData<NodeCommunity>(node);
+            ndata = nodeData.data;
+          },
+          katana::loopname("CDLP_Apply"));
+
+      applyBag.clear();
+      iterations += 1;
+    }
+    katana::ReportStatSingle("CDLP_Synchronous", "iterations", iterations);
+  }
+};
+
+struct CdlpAsynchronousAlgo {
+  using CommunityType = uint64_t;
+  struct NodeCommunity : public katana::PODProperty<CommunityType> {};
+
+  using NodeData = std::tuple<NodeCommunity>;
+  using EdgeData = std::tuple<>;
+  using Graph = katana::TypedPropertyGraphView<
+      katana::PropertyGraphViews::Undirected, NodeData, EdgeData>;
+  using GNode = typename Graph::Node;
+
+  CdlpPlan& plan_;
+  CdlpAsynchronousAlgo(CdlpPlan& plan) : plan_(plan) {}
+
+  void Initialize(Graph* graph) {
+    katana::do_all(katana::iterate(*graph), [&](const GNode& node) {
+      auto& ndata = graph->GetData<NodeCommunity>(node);
+      ndata = node;
+    });
+  }
+
+  void Deallocate(Graph*) {}
+
+  void operator()(Graph*, size_t) {}
+};
+
+}  //namespace
+
+template <typename Algorithm>
+static katana::Result<void>
+CdlpWithWrap(
+    katana::PropertyGraph* pg, std::string output_property_name, CdlpPlan plan,
+    size_t max_iterations) {
+  katana::EnsurePreallocated(
+      2,
+      pg->topology().num_nodes() * sizeof(typename Algorithm::NodeCommunity));
+  katana::ReportPageAllocGuard page_alloc;
+
+  if (auto r = ConstructNodeProperties<
+          std::tuple<typename Algorithm::NodeCommunity>>(
+          pg, {output_property_name});
+      !r) {
+    return r.error();
+  }
+  auto pg_result = Algorithm::Graph::Make(pg, {output_property_name}, {});
+  if (!pg_result) {
+    return pg_result.error();
+  }
+  auto graph = pg_result.value();
+
+  Algorithm algo(plan);
+
+  algo.Initialize(&graph);
+
+  katana::StatTimer execTime("CDLP");
+
+  execTime.start();
+  algo(&graph, max_iterations);
+  execTime.stop();
+
+  algo.Deallocate(&graph);
+  return katana::ResultSuccess();
+}
+
+katana::Result<void>
+katana::analytics::Cdlp(
+    PropertyGraph* pg, const std::string& output_property_name,
+    size_t max_iterations, CdlpPlan plan) {
+  switch (plan.algorithm()) {
+  case CdlpPlan::kSynchronous:
+    return CdlpWithWrap<CdlpSynchronousAlgo>(
+        pg, output_property_name, plan, max_iterations);
+  /// TODO: Asynchronous Algorithm will be implemented later after Synchronous
+  /// is done for both shared and distributed versions.
+  /*
+  case CdlpPlan::kAsynchronous:
+    return CdlpWithWrap<CdlpAsynchronousAlgo>(
+        pg, output_property_name, plan, max_iterations);
+  */
+  default:
+    return ErrorCode::InvalidArgument;
+  }
+}
+
+katana::Result<CdlpStatistics>
+katana::analytics::CdlpStatistics::Compute(
```

Comment:
This function is now being used by louvain, cc, and cdlp, basically everything which is calculating communities. Is it possible to move it to some common .h file in libgalois/include/analytics to avoid code duplication.

Reply:
I agree we should do this. This function for cc and cdlp is exactly the same while for louvain is more complex but we can come up with a common function for them and put it in a common .h file.  though I think it should be done in another task/pr.